### DISCUSSION TITLE
Unknown blend mode - What do I need to change?

### DISCUSSION THREAD

**Discussion by LudoBermejoES** [ptrumpis/snap-camera-server#101] [Q&A] [answered]
Hi!

I'm getting the same problem other users found when creating a lens in Lens Studio. Here is the log:

Core: CORE: using complex effect with id lns://C:/Program Files/Snap Inc/Snap Camera/Resources/RetouchLens.bundle/lens.lns@@
Core: [LS::Graphics::Context] this - 000000C6DCAFB6E0, glContext - 0000000000030001, glContextMode - 2
Core: [SGRenderer] Experiment value: 0
Core: bindScript executed
Core: [engine] engineComplexEffect started loading lens with id -1
Core: [engine] The lens under development mode is OFF
Core: [OpenGL] contextsMap doesn't contain eglContext 0000000000030001
Core: [LS::Graphics::~Context] this - 000000C6DCAFB6E0, glContext - 0000000000030001, glContextMode - 2
Core: [LS::Graphics::Context] this - 000000C6DCAFB460, glContext - 0000000000030001, glContextMode - 2
Core: [OpenGL] contextsMap doesn't contain eglContext 0000000000030001
Core: [LS::Graphics::~Context] this - 000000C6DCAFB460, glContext - 0000000000030001, glContextMode - 2
Core: Tracking requirements changed from 256 to 259 in context 0000025D6E9E5298
Core: [GenericPool] Created new LS::engine::Rendering::GLTextureLegacy. If you see this message too often, you probably allocating too many resources per frame.
Core: LS::engine::Infrastructure::engineComplexEffect LensID: -1, load + turnOn + 1 frame: 0.094704
Core: [GenericPool] Created new LS::engine::Rendering::GLTextureLegacy. If you see this message too often, you probably allocating too many resources per frame.
Core: [GenericPool] Created new LS::engine::Rendering::GLTextureLegacy. If you see this message too often, you probably allocating too many resources per frame.
Core: LS::engine::Infrastructure::engineComplexEffect LensID: -1, load + turnOn + 30 frames: 1.019386
VCam::ActionType::SetActiveLensId QVariant(VCam::LensActivationEvent, )
VCam::ActionType::ActiveLensChanged QVariant(VCam::LensInfo, )
VCam::ActionType::SetCoreProcessingLens QVariant(VCam::CoreProcessingLens, )
VCam::ActionType::SetMainLensHint QVariant(QString, "")
VCam::ActionType::LensDownloaded QVariant(QString, "63544770992")
VCam::ActionType::SetCoreProcessingOverlay QVariant(QImage, QImage(null))
Core: Snappables: no active session to send
Core: CORE: using complex effect with id lns://C:/Users/LudoB/AppData/Local/Snap/Snap Camera/cache/lenses/63544770992/lens.lns@@
Core: [LS::Graphics::Context] this - 000000C6DCAFB6E0, glContext - 0000000000030001, glContextMode - 2
Core: [SGRenderer] Experiment value: 0
Core: bindScript executed
Core: [engine] engineComplexEffect started loading lens with id 63544770992
Core: [engine] The lens under development mode is OFF
Core: [OpenGL] contextsMap doesn't contain eglContext 0000000000030001
Core: [LS::Graphics::~Context] this - 000000C6DCAFB6E0, glContext - 0000000000030001, glContextMode - 2
Core: [LS::Graphics::Context] this - 000000C6DCAFB460, glContext - 0000000000030001, glContextMode - 2
Core: Unknown blend mode 
Core: [engine] engine exception while processing lens with id 63544770992
Core: [OpenGL] contextsMap doesn't contain eglContext 0000000000030001
Core: [LS::Graphics::~Context] this - 000000C6DCAFB460, glContext - 0000000000030001, glContextMode - 2
"Lens not loaded exception: Unknown blend mode"
Core: Tracking requirements changed from 259 to 256 in context 0000025D6E9E5298
qt.qpa.mime: Retrying to obtain clipboard.

And here are the blendmodes I found in the XML:

          <int32 key="blendmode">16</int32>
          <int32 key="blendmode">6</int32>
          
I know you told you could fix it by changing the blendmode, but I'm not sure what values are needed. 

Would you be so kind to help me with that?

Thanks!

**Comment by ptrumpis** [comment, marked as answer]
```
<int32 key="blendmode">16</int32>
```

Blendmode 16 is the unsupported one by snap camera.
You can change the value to any number between 0 and 12.

### Reference Table
BlendMode | Value | Expression
-- | -- | --
Normal | 0 | SrcAlpha, OneMinusSrcAlpha
MultiplyLegacy [DEPRECATED] | 1 | DstColor, OneMinusSrcAlpha
AddLegacy [DEPRECATED] | 2 | One, One
Screen | 3 | One, OneMinusSrcColor
PremultipliedAlpha | 4 | One, OneMinusSrcAlpha
AlphaToCoverage | 5 | Blend Disabled
Disabled | 6 | Blend Disabled
Add | 7 | SrcAlpha, One
AlphaTest | 8 | Blend Disabled
ColoredGlass | 9 | Blend Disabled
Multiply | 10 | DstColor, Zero
Min | 11 | One, One
Max | 12 | One, One

13 might also be supported but I don't know what it stands for yet.

**Reply by LudoBermejoES** [reply]
Great, thanks! I'll try right now!

**Comment by LudoBermejoES** [comment]
After chaning blend mode to 0, I got another error:

"Lens not loaded exception: Effect with id: 53884280875 is malformed or empty - no proper configuration found in lns://C:/Users/LudoB/AppData/Local/Snap/Snap Camera/cache/lenses/53884280875/lens.lns@@"

That's weird. I tried with zip, I tried with lens, both with the same results.

If you are interested, here is the zip: https://www.transfernow.net/dl/2024020312tvOg0j

And the original one is https://www.snapchat.com/unlock/?type=SNAPCODE&uuid=8aa6f158825e431bb078d92513525870&metadata=01

**Reply by ptrumpis** [reply]
Did you use Snap Lens Tool to unpack and re-pack the lens?

I patched the lens if you want to download
[UnzipMe.zip](https://github.com/ptrumpis/snap-camera-server/files/14151428/UnzipMe.zip)

Blendmode 0 seems to create a black background for me.
While blendmode 3 seems to fix the lens.
Sometimes you have to play around with the values.

![demon](https://github.com/ptrumpis/snap-camera-server/assets/116500225/3fa8198e-96c7-4b7c-8945-bc3d6a19bf9d)

**Reply by LudoBermejoES** [reply]
Ooooh great, it seems it works! Fantastic! Thanks!
